org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacProblemConverter.java

```diff
@@ -26,33 +26,84 @@
 import com.sun.tools.javac.code.Kinds;
 import com.sun.tools.javac.code.Symbol;
 import com.sun.tools.javac.tree.JCTree.JCClassDecl;
+import com.sun.tools.javac.tree.JCTree.JCVariableDecl;
 import com.sun.tools.javac.util.DiagnosticSource;
 import com.sun.tools.javac.util.JCDiagnostic;
 import com.sun.tools.javac.util.Position;
 
 public class JavacProblemConverter {
+	private static final String COMPILER_WARN_NON_SERIALIZABLE_INSTANCE_FIELD = "compiler.warn.non.serializable.instance.field";
+	private static final String COMPILER_WARN_MISSING_SVUID = "compiler.warn.missing.SVUID";
+
 	public static JavacProblem createJavacProblem(Diagnostic<? extends JavaFileObject> diagnostic) {
 		int problemId = toProblemId(diagnostic);
-		return switch (problemId) {
-			case IProblem.MissingSerialVersion -> classLevelProblem(diagnostic, problemId);
-			default -> defaultProblem(diagnostic, problemId);
-		};
-	}
-
-	private static JavacProblem defaultProblem(Diagnostic<? extends JavaFileObject> diagnostic, int problemId) {
+		int[] diagnosticPosition = getDiagnosticPosition(diagnostic);
 		return new JavacProblem(
 				diagnostic.getSource().getName().toCharArray(),
 				diagnostic.getMessage(Locale.getDefault()),
 				diagnostic.getCode(),
 				problemId,
 				new String[0],
 				toSeverity(diagnostic),
-				(int) Math.min(diagnostic.getPosition(), diagnostic.getStartPosition()),
-				(int) (diagnostic.getEndPosition() - 1),
+				diagnosticPosition[0],
+				diagnosticPosition[1],
 				(int) diagnostic.getLineNumber(),
 				(int) diagnostic.getColumnNumber());
 	}
 
+	// result[0] - startPosition
+	// result[1] - endPosition
+	private static int[] getDiagnosticPosition(Diagnostic<? extends JavaFileObject> diagnostic) {
+		if (diagnostic != null) {
+			switch (diagnostic.getCode()) {
+			case COMPILER_WARN_MISSING_SVUID: {
```

Comment:
I feel like we shouldn't test diagnostic codes (there's over a 1000 errors IIRC), but simply rely on `jcDiagnostic.getDiagnosticPosition() instanceof JCClassDecl` and `jcDiagnostic.getDiagnosticPosition() instanceof JCVariableDecl`. Unless there are cases where this would not apply

Reply:
Fixed.